pkg/scheduler/framework/plugins/nodevolumelimits/csi.go

```diff
@@ -72,13 +76,68 @@ func (pl *CSILimits) Name() string {
 	return CSIName
 }
 
-// EventsToRegister returns the possible events that may make a Pod
+// EventsToRegister returns the possible events that may make a Pod.
 // failed by this plugin schedulable.
 func (pl *CSILimits) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
 		{Event: framework.ClusterEvent{Resource: framework.CSINode, ActionType: framework.Add}},
-		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}},
+	}
+}
+
+func (pl *CSILimits) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	deletedPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		return framework.Queue, fmt.Errorf("unexpected objects in isSchedulableAfterPodDeleted: %w", err)
+	}
+
+	if len(deletedPod.Spec.Volumes) == 0 {
+		return framework.QueueSkip, nil
+	}
+
+	if deletedPod.Spec.NodeName == "" {
+		return framework.QueueSkip, nil
+	}
+
+	csiNode, err := pl.csiNodeLister.Get(deletedPod.Spec.NodeName)
+	if apierrors.IsNotFound(err) {
+		// When csiNode for a deleted Pod is not found, it doesn't make the pod schedulable
+		// because csiNode should be recreated again to let the pod go to this node.
+		logger.V(5).Info("A PVC for the Pod is not found, this Pod won't be schedulable until it's recreated", "pod", klog.KObj(pod), "namespace", pod.Namespace)
+		return framework.QueueSkip, nil
+	} else if err != nil {
+		return framework.Queue, fmt.Errorf("could not get a CSINode object: %w", err)
 	}
+
+	deletedPodAttachedVolumes := make(map[string]string)
+	if err := pl.filterAttachableVolumes(logger, deletedPod, csiNode, false /* existing pod */, deletedPodAttachedVolumes); err != nil {
+		return framework.Queue, fmt.Errorf("failed to filter attachable volumes from %v: %w", klog.KObj(deletedPod), err)
+	}
+	deletedPodDrivers := sets.Set[string]{}
+	for _, driverName := range deletedPodAttachedVolumes {
+		deletedPodDrivers.Insert(driverName)
+	}
+
+	schedulingPodAttachedVolumes := make(map[string]string)
+	if err := pl.filterAttachableVolumes(logger, pod, csiNode, false /* new pod */, schedulingPodAttachedVolumes); err != nil {
```

Comment:
```suggestion
	if err := pl.filterAttachableVolumes(logger, pod, csiNode, true /* new pod */, schedulingPodAttachedVolumes); err != nil {
```

right?

Reply:
It means https://github.com/kubernetes/kubernetes/pull/121952 should be merged before merging this PR.

Reply:
> It means #121952 should be merged before merging this PR.

@sanposhiho 
Is this a comment on https://github.com/kubernetes/kubernetes/pull/121508#discussion_r1451667884?

Reply:
Sorry, lack of words. 
If we change here to `true`, `filterAttachableVolumes` returns an error when PVC not found. After #121952, we can catch PVC-not-found error here and we should return `QueueSkip` if the cluster doesn't have a required PVC of `pod` yet. (#121952 also adds PVC/Added event to `EventsToRegister`.)

```go
if err := pl.filterAttachableVolumes(logger, pod, csiNode, true /* new pod */, schedulingPodAttachedVolumes); err != nil {
  if apierrors.IsNotFound(err) {
    // pod requires non-existing PVC - pod will never be schedulable until PVC is created.
    // We return QueueSkip here and catch the moment that a required PVC is created via PVC/Added event.
    return framework.QueueSkip, nil
  }
}
```

Reply:
👍

Reply:
https://github.com/kubernetes/kubernetes/pull/121508/commits/a83939eaf881ba5a8abfae92456e762ff92ccc99

Reply:
https://github.com/kubernetes/kubernetes/pull/121952 is merged, you should be write the test case once the branch is rebased.

Reply:
Added
https://github.com/kubernetes/kubernetes/pull/121508/commits/e2af012e62ebffff42f41d3c01df1e181380aea1